api/incubator/src/main/java/io/opentelemetry/api/incubator/propagation/CaseInsensitiveMap.java

```diff
@@ -8,24 +8,38 @@
 import java.util.HashMap;
 import java.util.Locale;
 import java.util.Map;
+import javax.annotation.Nonnull;
 import javax.annotation.Nullable;
 
 class CaseInsensitiveMap extends HashMap<String, String> {
 
   private static final long serialVersionUID = -4202518750189126871L;
 
+  CaseInsensitiveMap() {}
+
   CaseInsensitiveMap(Map<String, String> carrier) {
-    super(carrier);
+    if (carrier != null) {
+      this.putAll(carrier);
+    }
   }
 
   @Override
   public String put(String key, String value) {
-    return super.put(key.toLowerCase(Locale.ROOT), value);
+    return super.put(getKeyLowerCase(key), value);
+  }
+
+  @Override
+  public void putAll(Map<? extends String, ? extends String> m) {
+    m.forEach(this::put);
+  }
+
+  private static String getKeyLowerCase(@Nonnull String key) {
```

Comment:
I think we normally don't include NonNull annotations?

Reply:
hmm. I don't remember either way. Happy  to remove if you think we should skip it.

Reply:
done